src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -215,17 +215,54 @@ private static unsafe nuint GetIndexOfFirstNonAsciiByte_Default(byte* pBuffer, n
             goto Finish;
         }
 
-        private static unsafe nuint GetIndexOfFirstNonAsciiByte_Sse2(byte* pBuffer, nuint bufferLength)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static ulong GetMaskOfMostSignificantBits(Vector128<byte> value)
+        {
+            Debug.Assert(BitConverter.IsLittleEndian);
+            if (!AdvSimd.Arm64.IsSupported)
+            {
+                throw new PlatformNotSupportedException();
+            }
+
+            // extractedBits[i] = (value[i] & 0x80 == 0x80) & (1 << (12 * (i % 2)))
+            Vector128<byte> s_mostSignficantBitMask = Vector128.Create((byte)0x80);
+            Vector128<byte> s_bitmask = Vector128.Create((ushort)0x1001).AsByte();
+            Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareTest(value, s_mostSignficantBitMask);
+            Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitmask);
+
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
+            return extractedBits.AsUInt64().ToScalar();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ContainsNonAsciiByte(Vector128<ushort> value)
```

Comment:
>ContainsNonAsciiByte [](start = 27, length = 20)

This method should not contain `Byte` because it doesn't operate on `byte` but `ushort`.

Reply:
Perhaps, you just need one method with byte overload and use .AsByte() for arrays with ushort elements.

Reply:
Hhm, is that true. I think the intent is to check if bit 15 in each lane is set or not (the caller is operating on utf16). So imagine a data where all lanes are `<506, 506, 506, 506, 506, 506, 506, 506>` but when type casted to `.AsByte()` they will be `<250, 1, 250, 1, 250, 1, 250, 1, 250, 1, 250, 1, 250, 1, 250, 1>` and return that it contains non-ascii data but in fact it doesn't.

Reply:
@kunalspathak My understanding was that it's just a convenience method to do the same kind of check as done for bytes. @pgovind Can you please confirm what needs to happen here?

If Kunal describes is correct - then I agree that the name shouldn't be *Bytes

Reply:
Name shouldn't be bytes. Changed the name and the bit pattern I am `&`ing with